src/workerd/util/state-machine.h

```diff
@@ -0,0 +1,3897 @@
+  //
+  // WARNING: Bypasses terminal state protection. See forceTransitionTo() docs
+  // for when this is appropriate vs. suspicious.
+  template <typename... Args>
+  ErrorState& forceError(Args&&... args) KJ_LIFETIMEBOUND {
+    this->requireUnlocked();
+    return this->state.template init<ErrorState>(kj::fwd<Args>(args)...);
+  }
+
+  // Generic transition (for backward compatibility)
+  template <typename S, typename... Args>
+  S& transitionTo(Args&&... args) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<S, ActiveState, ClosedState, ErrorState>)
+  {
+    this->requireUnlocked();
+    return this->state.template init<S>(kj::fwd<Args>(args)...);
+  }
+};
+
+// =============================================================================
+// Validated State Machine
+// =============================================================================
+
+// A state machine with compile-time transition validation.
+// The TransitionPolicy must have a static method:
+//   template <typename From, typename To>
+//   static constexpr bool isAllowed();
+//
+// Invalid transitions will cause a compile-time error.
+
+template <typename TransitionPolicy, typename... States>
+class ValidatedStateMachine: public StateMachine<States...> {
+  using Base = StateMachine<States...>;
+
+ public:
+  using Base::Base;
+  using Base::get;
+  using Base::is;
+  using Base::tryGet;
+
+  // Unvalidated transition (same as base)
+  template <typename To, typename... Args>
+  To& transitionTo(Args&&... args) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<To, States...>)
+  {
+    this->requireUnlocked();
+    return this->state.template init<To>(kj::fwd<Args>(args)...);
+  }
+
+  // Validated transition from a specific state.
+  // Compile-time error if From -> To is not allowed by the policy.
+  template <typename From, typename To, typename... Args>
+  To& checkedTransitionFromTo(Args&&... args) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<From, States...>) && (_::isOneOf<To, States...>) &&
+      (TransitionPolicy::template isAllowed<From, To>())
+  {
+    this->requireUnlocked();
+    KJ_REQUIRE(this->template is<From>(),
+        "State machine transition precondition failed: not in expected state");
+    return this->state.template init<To>(kj::fwd<Args>(args)...);
+  }
+
+  // Try validated transition - returns none if not in From state.
+  template <typename From, typename To, typename... Args>
+  KJ_WARN_UNUSED_RESULT kj::Maybe<To&> tryCheckedTransitionFromTo(Args&&... args) KJ_LIFETIMEBOUND
+    requires(_::isOneOf<From, States...>) && (_::isOneOf<To, States...>) &&
+      (TransitionPolicy::template isAllowed<From, To>())
+  {
+    this->requireUnlocked();
+    if (this->template is<From>()) {
+      return this->state.template init<To>(kj::fwd<Args>(args)...);
+    }
+    return kj::none;
+  }
+};
+
+// =============================================================================
+// Observable State Machine
```

Comment:
Unused, and not mentioned in this header's doc comment

Reply:
Will remove.